execution/tvf/percentile_watermark_generator.go

```diff
@@ -0,0 +1,297 @@
+package tvf
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/execution"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+type PercentileWatermarkGenerator struct {
+	source     execution.Node
+	timeField  octosql.VariableName
+	events     execution.Expression
+	percentile execution.Expression
+	frequency  execution.Expression
+}
+
+func NewPercentileWatermarkGenerator(source execution.Node, timeField octosql.VariableName, events, percentile, frequency execution.Expression) *PercentileWatermarkGenerator {
+	return &PercentileWatermarkGenerator{
+		source:     source,
+		timeField:  timeField,
+		events:     events,
+		percentile: percentile,
+		frequency:  frequency,
+	}
+}
+
+func (w *PercentileWatermarkGenerator) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, *execution.ExecOutput, error) {
+	tx := storage.GetStateTransactionFromContext(ctx)
+	sourceStreamID, err := execution.GetSourceStreamID(tx.WithPrefix(streamID.AsPrefix()), octosql.MakePhantom())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source stream ID")
+	}
+
+	source, _, err := w.source.Get(ctx, variables, sourceStreamID) // we don't need execOutput here since we become new watermark source
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source")
+	}
+
+	events, err := w.events.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark events")
+	}
+	if events.GetType() != octosql.TypeInt || events.AsInt() < 1 {
+		return nil, nil, errors.Errorf("invalid watermark events: %v", events)
+	}
+
+	percentile, err := w.percentile.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark percentile")
+	}
+	if percentile.GetType() != octosql.TypeInt || percentile.AsInt() < 1 || percentile.AsInt() > 99 {
```

Comment:
I think maybe it looks just a bit better when values such as this 1 and 99 are named. Maybe add constants 
`minPercentileValue` and `maxPercentileValue`? (I'm open for this discussion about this one tho)

Reply:
Or maybe this is an overkill - I dunno

Reply:
I personally think this would be an overkill, what about you, @cube2222 ?